In [27]:
import csv
import datetime

In [17]:
data = list(csv.reader(open('guns.csv')))

In [18]:
#First 5 rows with header
data[0:5]

[['',
  'year',
  'month',
  'intent',
  'police',
  'sex',
  'age',
  'race',
  'hispanic',
  'place',
  'education'],
 ['1',
  '2012',
  '01',
  'Suicide',
  '0',
  'M',
  '34',
  'Asian/Pacific Islander',
  '100',
  'Home',
  '4'],
 ['2', '2012', '01', 'Suicide', '0', 'F', '21', 'White', '100', 'Street', '3'],
 ['3',
  '2012',
  '01',
  'Suicide',
  '0',
  'M',
  '60',
  'White',
  '100',
  'Other specified',
  '4'],
 ['4', '2012', '02', 'Suicide', '0', 'M', '64', 'White', '100', 'Home', '4']]

In [19]:
headers = data[0]
data = data[1:]

In [20]:
headers

['',
 'year',
 'month',
 'intent',
 'police',
 'sex',
 'age',
 'race',
 'hispanic',
 'place',
 'education']

In [21]:
#First 5 rows without header
data[0:5]

[['1',
  '2012',
  '01',
  'Suicide',
  '0',
  'M',
  '34',
  'Asian/Pacific Islander',
  '100',
  'Home',
  '4'],
 ['2', '2012', '01', 'Suicide', '0', 'F', '21', 'White', '100', 'Street', '3'],
 ['3',
  '2012',
  '01',
  'Suicide',
  '0',
  'M',
  '60',
  'White',
  '100',
  'Other specified',
  '4'],
 ['4', '2012', '02', 'Suicide', '0', 'M', '64', 'White', '100', 'Home', '4'],
 ['5',
  '2012',
  '02',
  'Suicide',
  '0',
  'M',
  '31',
  'White',
  '100',
  'Other specified',
  '2']]

In [22]:
#No of deaths each year
years_counts = {}
for row in data:
    year = row[1]
    if year in years_counts:
        years_counts[year] += 1
    else:
        years_counts[year] = 1

In [23]:
years_counts

{'2012': 33563, '2013': 33636, '2014': 33599}

In [29]:
dates = [datetime.datetime(year=int(row[1]),month=int(row[2]),day = 1) for row in data]

In [31]:
dates[0:5]

[datetime.datetime(2012, 1, 1, 0, 0),
 datetime.datetime(2012, 1, 1, 0, 0),
 datetime.datetime(2012, 1, 1, 0, 0),
 datetime.datetime(2012, 2, 1, 0, 0),
 datetime.datetime(2012, 2, 1, 0, 0)]

In [35]:
date_counts = {'2012':{},'2013':{},'2014':{}}
for row in dates:
    year = str(row.year)
    month = str(row.month)
    if month in date_counts[year]:
        date_counts[year][month] += 1
    else:
        date_counts[year][month] = 1


In [36]:
date_counts

{'2012': {'1': 2758,
  '10': 2733,
  '11': 2729,
  '12': 2791,
  '2': 2357,
  '3': 2743,
  '4': 2795,
  '5': 2999,
  '6': 2826,
  '7': 3026,
  '8': 2954,
  '9': 2852},
 '2013': {'1': 2864,
  '10': 2808,
  '11': 2758,
  '12': 2765,
  '2': 2375,
  '3': 2862,
  '4': 2798,
  '5': 2806,
  '6': 2920,
  '7': 3079,
  '8': 2859,
  '9': 2742},
 '2014': {'1': 2651,
  '10': 2865,
  '11': 2756,
  '12': 2857,
  '2': 2361,
  '3': 2684,
  '4': 2862,
  '5': 2864,
  '6': 2931,
  '7': 2884,
  '8': 2970,
  '9': 2914}}

In [37]:
sex_counts = {}
race_counts = {}
for row in data:
    sex = row[5]
    race = row[7]
    if sex in sex_counts:
        sex_counts[sex]+= 1
    else:
        sex_counts[sex] = 1
    if race in race_counts:
        race_counts[race]+= 1
    else:
        race_counts[race]= 1

In [38]:
sex_counts

{'F': 14449, 'M': 86349}

In [39]:
race_counts

{'Asian/Pacific Islander': 1326,
 'Black': 23296,
 'Hispanic': 9022,
 'Native American/Native Alaskan': 917,
 'White': 66237}

## Majority of gun deaths are Males

### 'M': 86349 (Maximum)

#### We need to further find counts by education and age of the persons with gun death
#### Need population distribution to compare the counts of people by race

In [43]:
census = list(csv.reader(open('census.csv')))

In [44]:
census

[['Id',
  'Year',
  'Id',
  'Sex',
  'Id',
  'Hispanic Origin',
  'Id',
  'Id2',
  'Geography',
  'Total',
  'Race Alone - White',
  'Race Alone - Hispanic',
  'Race Alone - Black or African American',
  'Race Alone - American Indian and Alaska Native',
  'Race Alone - Asian',
  'Race Alone - Native Hawaiian and Other Pacific Islander',
  'Two or More Races'],
 ['cen42010',
  'April 1, 2010 Census',
  'totsex',
  'Both Sexes',
  'tothisp',
  'Total',
  '0100000US',
  '',
  'United States',
  '308745538',
  '197318956',
  '44618105',
  '40250635',
  '3739506',
  '15159516',
  '674625',
  '6984195']]

In [49]:
census[0]

['Id',
 'Year',
 'Id',
 'Sex',
 'Id',
 'Hispanic Origin',
 'Id',
 'Id2',
 'Geography',
 'Total',
 'Race Alone - White',
 'Race Alone - Hispanic',
 'Race Alone - Black or African American',
 'Race Alone - American Indian and Alaska Native',
 'Race Alone - Asian',
 'Race Alone - Native Hawaiian and Other Pacific Islander',
 'Two or More Races']

In [51]:
census_race_index = {}
for index,header in enumerate(list(census[0])):
    if not(header in census_race_index):
        census_race_index[header] = index
        

In [52]:
census_race_index

{'Geography': 8,
 'Hispanic Origin': 5,
 'Id': 0,
 'Id2': 7,
 'Race Alone - American Indian and Alaska Native': 13,
 'Race Alone - Asian': 14,
 'Race Alone - Black or African American': 12,
 'Race Alone - Hispanic': 11,
 'Race Alone - Native Hawaiian and Other Pacific Islander': 15,
 'Race Alone - White': 10,
 'Sex': 3,
 'Total': 9,
 'Two or More Races': 16,
 'Year': 1}

In [56]:
race_mapping = {'Asian/Pacific Islander' : ['Race Alone - Asian','Race Alone - Native Hawaiian and Other Pacific Islander'],  'Black' : ['Race Alone - Black or African American'],'Hispanic' : ['Race Alone - Hispanic'],'Native American/Native Alaskan' :['Race Alone - American Indian and Alaska Native'],'White' : ['Race Alone - White']}


In [57]:
race_mapping

{'Asian/Pacific Islander': ['Race Alone - Asian',
  'Race Alone - Native Hawaiian and Other Pacific Islander'],
 'Black': ['Race Alone - Black or African American'],
 'Hispanic': ['Race Alone - Hispanic'],
 'Native American/Native Alaskan': ['Race Alone - American Indian and Alaska Native'],
 'White': ['Race Alone - White']}

In [62]:
race_per_hundredk= {}
for race,value in race_counts.items():
    mapping = race_mapping[race]
    total = 0
    for i in mapping:
        total = total + int(census[1][census_race_index[i]])
    per_hundredk = (value / total) * 100000
    race_per_hundredk[race] = per_hundredk
    

In [63]:
race_per_hundredk

{'Asian/Pacific Islander': 8.374309664161762,
 'Black': 57.8773477735196,
 'Hispanic': 20.220491210910907,
 'Native American/Native Alaskan': 24.521955573811088,
 'White': 33.56849303419181}

### Now we can say that Gun death per hundredk for Black People is maximum and it is highest in Males
### So Male black people are highest in gun  deaths


In [64]:
intents = [row[3] for row in data]

In [65]:
races = [row[7] for row in data]

In [66]:
intents

['Suicide',
 'Suicide',
 'Suicide',
 'Suicide',
 'Suicide',
 'Suicide',
 'Undetermined',
 'Suicide',
 'Accidental',
 'Suicide',
 'Suicide',
 'Suicide',
 'Suicide',
 'Suicide',
 'Homicide',
 'Suicide',
 'Suicide',
 'Suicide',
 'Homicide',
 'Suicide',
 'Suicide',
 'Suicide',
 'Homicide',
 'Suicide',
 'Homicide',
 'Homicide',
 'Suicide',
 'Suicide',
 'Suicide',
 'Homicide',
 'Suicide',
 'Suicide',
 'Suicide',
 'Suicide',
 'Homicide',
 'Undetermined',
 'Suicide',
 'Suicide',
 'Suicide',
 'Suicide',
 'Suicide',
 'Homicide',
 'Suicide',
 'Suicide',
 'Suicide',
 'Suicide',
 'Suicide',
 'Homicide',
 'Homicide',
 'Homicide',
 'Suicide',
 'Suicide',
 'Suicide',
 'Undetermined',
 'Suicide',
 'Suicide',
 'Accidental',
 'Homicide',
 'Suicide',
 'Homicide',
 'Suicide',
 'Homicide',
 'Suicide',
 'Suicide',
 'Suicide',
 'Suicide',
 'Suicide',
 'Suicide',
 'Suicide',
 'Suicide',
 'Suicide',
 'Accidental',
 'Suicide',
 'Accidental',
 'Suicide',
 'Undetermined',
 'Suicide',
 'Suicide',
 'Suicide',
 'Homi

In [67]:
races

['Asian/Pacific Islander',
 'White',
 'White',
 'White',
 'White',
 'Native American/Native Alaskan',
 'White',
 'Native American/Native Alaskan',
 'White',
 'Black',
 'White',
 'Native American/Native Alaskan',
 'White',
 'Native American/Native Alaskan',
 'White',
 'Native American/Native Alaskan',
 'White',
 'Native American/Native Alaskan',
 'Asian/Pacific Islander',
 'White',
 'Native American/Native Alaskan',
 'White',
 'Black',
 'White',
 'White',
 'White',
 'Native American/Native Alaskan',
 'White',
 'White',
 'White',
 'Native American/Native Alaskan',
 'White',
 'Native American/Native Alaskan',
 'White',
 'White',
 'White',
 'Native American/Native Alaskan',
 'Native American/Native Alaskan',
 'White',
 'White',
 'Native American/Native Alaskan',
 'Asian/Pacific Islander',
 'White',
 'Native American/Native Alaskan',
 'Native American/Native Alaskan',
 'White',
 'White',
 'White',
 'Native American/Native Alaskan',
 'Asian/Pacific Islander',
 'White',
 'White',
 'Native Ame

In [72]:
#Why Using whole data to find this????
homicide_race_counts = {}
for row in data:
    intent = row[3]
    race = row[7]
    if intent == 'Homicide':
        if race in homicide_race_counts :
            homicide_race_counts[race]+= 1
        else:
            homicide_race_counts[race]= 1

In [70]:
homicide_race_counts

{'Asian/Pacific Islander': 559,
 'Black': 19510,
 'Hispanic': 5634,
 'Native American/Native Alaskan': 326,
 'White': 9147}

In [73]:
#Modify the counts in place
for race,value in homicide_race_counts.items():
    mapping = race_mapping[race]
    total = 0
    for i in mapping:
        total = total + int(census[1][census_race_index[i]])
    per_hundredk = (value / total) * 100000
    homicide_race_counts[race] = per_hundredk
    

In [74]:
homicide_race_counts

{'Asian/Pacific Islander': 3.530346230970155,
 'Black': 48.471284987180944,
 'Hispanic': 12.627161104219914,
 'Native American/Native Alaskan': 8.717729026240365,
 'White': 4.6356417981453335}

#### Still we can see Black people in Homicide are maximum